### PR TITLE
Add skip connection

### LINKED ISSUE #330: Include previous ground-truth in the decoder
Add a UNet-style cross-link where we do a final Conv2D between yesterday's ground-truth and the output of the decoder

### PR DESCRIPTION
**Pull request by jemrobinson**
## Summary

Adds a skip-connection from the most recent target date (c.f. persistence) to every forecast step.

This allows the decode to predict the *residual* rather than generating the whole distribution from scratch. Three different skip options are implemented here (plus the "no-skip" option which reproduces the current behaviour).

- `additive` (add the persistence to the prediction)
- `convolutional` (convolve the persistence with the prediction)
- `gated` (output per-pixel weighted combination of persistence and prediction)

**TL;DR** Clearest single improvement of anything tried so far! All three skip-connection variants beat the no-skip-connection baseline for `cnn-unet-cnn` by around 10x. `gated` is the best of the four, `additive` a close second, `convolutional` a clear step behind both. This applies to both 2-day and 14-day rollout.

**Caveat:** sea-ice evolution over 14-days is still poor (and not very correlated to the real changes), it's just starting from a better baseline than previously.

Closes #330

## Details

- To get the skip-connection to work with multistage training we do the following:
  - `EncoderStage` has no skip-connection (or masking) on its decoder since it's reconstructing the input
  - `DecoderStage` now requires two timesteps, using yesterday's target as persistence
  - `ProcessorStage` follows the same logic as the full `EncodeProcessDecode` model
- Set decoder output behaviour: `range_restriction > skip_connection > mask` so that the persistence is applied at the same scale (normalised to [0,1]) as the model prediction.
- Added some tests

<details>
<summary>Decoder rollout diagram</summary>

```mermaid
flowchart TD
    X["decoder(x)"] --> RESTRICT["restrict_range (none/sigmoid/tanh/clamp)"]
    RESTRICT --> SKIP{"skip_connection configured?"}
    PERSIST["persistence\n(most recent target frame)"] --> SKIP
    SKIP -->|"yes"| COMBINE["SkipConnection(main, skip)"]
    COMBINE --> CLAMP["clamp(0, 1)"]
    SKIP -->|"no"| CLAMP2["(unchanged)"]
    CLAMP --> MASK["mask (zero inactive/land cells)"]
    CLAMP2 --> MASK
    MASK --> OUT["output"]

    classDef terminal fill:#e8e8e8,stroke:#888,color:#111
    classDef conv fill:#cde4ff,stroke:#4a77bd,color:#111
    classDef decision fill:#ffe4b5,stroke:#cc8400,color:#111

    class X,OUT,PERSIST terminal
    class RESTRICT,COMBINE,CLAMP,CLAMP2,MASK conv
    class SKIP decision
```
</details>

## Comparison

Ran a skip-connection comparison on `cnn-unet-cnn` from the same commit (`e3e69bc`):

- Config: `baseline/02_cnn_unet_cnn --multistage`, `random=deterministic`
- Data: south hemisphere, 1y sample (train=2017 + 2018 excl. Jan, validate=2018-01, test=2019-01)
- Only the skip connection method varied: `++model.decoder.skip_connection.method={additive,convolutional,gated,none}`
- JASMIN (orchid partition), 1 GPU+8 CPU each

### Training (finetuning after multistage)

| Variant | Forecast days | Wall time | Validation loss | Validation rmse | Validation sieerror (km2) |
|---|---|---|---|---|---|
| `additive` | 2 | 5h37m22s | **0.000115** | **0.0149** | **204,086** |
| `gated` | 2 | 5h39m51s | 0.000127 | 0.0157 | 171,968 |
| `convolutional` | 2 | 5h15m56s | 0.000197 | 0.0198 | 236,435 |
| `none` | 2 | 5h56m18s | 0.001594 | 0.0569 | 1,027,176 |
| `additive` | 14 | 10h52m55s | 0.000891 | 0.0405 | 1,118,185 |
| `gated` | 14 | 8h49m50s | **0.000715** | **0.0368** | 734,616 |
| `convolutional` | 14 | 7h35m 8s | 0.000758 | 0.0376 | 1,181,265 |
| `none` | 14 | 10h44m24s | 0.001803 | 0.0607 | **447,265** |

### Evaluation

| Variant | Forecast days | test loss | test mae | test rmse | test accuracy | test sieerror (km2) |
|---|---|---|---|---|---|---|
| `additive` | 2 | **0.000112** | **0.00203** | 0.0147 | **99.79** | **123,275** |
| `gated` | 2 | 0.000123 | 0.00334 | **0.0155** | 99.79 | 132,234 |
| `convolutional` | 2 | 0.000438 | 0.00432 | 0.0293 | 99.76 | 194,097 |
| `none` | 2 | 0.001801 | 0.02016 | 0.0593 | 98.26 | 1,829,387 |
| `additive` | 14 | 0.000761 | 0.00871 | 0.0382 | 99.28 | 705,604 |
| `gated` | 14 | **0.000616** | **0.00597** | **0.0341** | **99.40** | **419,708** |
| `convolutional` | 14 | 0.000753 | 0.00635 | 0.0382 | 99.17 | 876,813 |
| `none` | 14 | 0.002101 | 0.01506 | 0.0657 | 98.46 | 792,214

`gated` is tied with `additive` at 2-day forecast and clearly first at 14-day. `convolutional` is a step behind both, and `none` decisively worst across every metric.

### Additive
media_images_output_static_2019-01-16-sea-ice-concentration_2_c25cf7e5cdf0ae509085

### Convolutional
media_images_output_static_2019-01-16-sea-ice-concentration_2_741c079298eae343d5ff

### Gated
media_images_output_static_2019-01-16-sea-ice-concentration_2_295967bdf14e7cafa080

### None
media_images_output_static_2019-01-16-sea-ice-concentration_2_87f3f3a9bc2853967551

<details>
<summary><b>W&B run links</b></summary>

| Run | Purpose | Link |
|---|---|---|
| bright-cherry-75 | `additive` training (2-day forecast) | https://wandb.ai/turing-seaice/train/runs/t3lxfklu |
| stellar-bird-395 | `additive` evaluate (2-day forecast) | https://wandb.ai/turing-seaice/evaluate/runs/5kbsol2l |
| astral-shadow-78 | `convolutional` training (2-day forecast) | https://wandb.ai/turing-seaice/train/runs/szs1hhyt |
| misunderstood-shape-396 | `convolutional` evaluate (2-day forecast) | https://wandb.ai/turing-seaice/evaluate/runs/9r9qzlxw |
| light-bush-76 | `gated` training (2-day forecast) | https://wandb.ai/turing-seaice/train/runs/3txmvsov |
| hardy-frost-397 | `gated` evaluate (2-day forecast) | https://wandb.ai/turing-seaice/evaluate/runs/s743xnix |
| eager-bee-79 | `none` training (2-day forecast) | https://wandb.ai/turing-seaice/train/runs/d1qcqpgs |
| worldly-tree-398 | `none` evaluate (2-day forecast) | https://wandb.ai/turing-seaice/evaluate/runs/6rnutliz |
| vital-silence-80 | `additive` training (14-day forecast) | https://wandb.ai/turing-seaice/train/runs/b11j0pge |
| treasured-rain-401 | `additive` evaluate (14-day forecast) | https://wandb.ai/turing-seaice/evaluate/runs/8zrbyyee |
| fearless-water-81 | `convolutional` training (14-day forecast) | https://wandb.ai/turing-seaice/train/runs/199s3ep9 |
| still-salad-399 | `convolutional` evaluate (14-day forecast) | https://wandb.ai/turing-seaice/evaluate/runs/ebb71msh |
| rose-leaf-82 | `gated` training (14-day forecast) | https://wandb.ai/turing-seaice/train/runs/lh76tvj7 |
| woven-cosmos-402 | `gated` evaluate (14-day forecast) | https://wandb.ai/turing-seaice/evaluate/runs/kcxflybd |
| dulcet-feather-83 | `none` training (14-day forecast) | https://wandb.ai/turing-seaice/train/runs/m0dxvhdv |
| elated-resonance-400 | `none` evaluate (14-day forecast) | https://wandb.ai/turing-seaice/evaluate/runs/vg5wax0n |

</details>

### CODE REVIEW

File: icenet_mp/models/decoders/base_decoder.py [hunk 1]
```diff
@@ -56,31 +80,20 @@ def forward(self, x: TensorNCHW) -> TensorNCHW:
...
-        regardless of bounding (applying it before would leak e.g. sigmoid(0)=0.5 into
-        masked cells). Called once by `rollout` after the per-frame `forward`, so every
-        decoder gets it automatically without having to call it themselves.
-
-        RangeRestriction choices are: none/sigmoid/tanh/clamp
-        """
-        return self.mask(self.restrict(x))
-
-    def rollout(self, x: TensorNTCHW) -> TensorNTCHW:
+    def rollout(self, x: TensorNTCHW, persistence: TensorNTCHW | None) -> TensorNTCHW:
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/decoders/base_decoder.py:128]: We want this to be a breaking change so that persistence is applied/not-applied deliberately. Docstring fixed in 784748f9.

File: icenet_mp/models/decoders/base_decoder.py [hunk 2]
```diff
@@ -90,5 +103,29 @@ def rollout(self, x: TensorNTCHW) -> TensorNTCHW:
...
         batch_size, n_timeslices = x.shape[0], x.shape[1]
-        output = self.finalise(self(x.reshape(-1, *self.data_space_in.chw)))
+        # Pass the latents through the decoder to return to output space
+        unbounded_output_nchw: TensorNCHW = self(x.reshape(-1, *self.data_space_in.chw))
+        # Apply range restriction to bring the output into the required range
+        bounded_output_nchw: TensorNCHW = self.restrict(unbounded_output_nchw)
+
+        if self.skip_connection:
+            if persistence is None:
+                msg = f"Decoder has skip_connection={self.skip_connection.method} but no persistence input was provided."
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/decoders/base_decoder.py]: Done in 0cb4ebf6

File: icenet_mp/models/multistage/processor_stage.py [hunk 3]
```diff
@@ -98,15 +99,20 @@ def encode_inputs(self, inputs: dict[str, TensorNTCHW]) -> TensorNTCHW:
...
         - decode with frozen decoder.rollout() -> output space NTCHW
         """
         combined_latent: TensorNTCHW = self.encode_inputs(inputs)
-        return self.decoder.rollout(self.processor.rollout(combined_latent).prediction)
+        processed = self.processor.rollout(combined_latent).prediction
+        # Add persistence skip connection if requested
+        persistence: TensorNTCHW | None = inputs[self.output_space.name][
+            :, -1, self.target_variable_indices, :, :
+        ].unsqueeze(1)
+        return self.decoder.rollout(processed, persistence)
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/multistage/processor_stage.py]: Done in b1dfe27e

File: icenet_mp/models/encode_process_decode.py [hunk 4]
```diff
@@ -133,8 +146,13 @@ def forward(self, inputs: dict[str, TensorNTCHW]) -> TensorNTCHW:
             latent_input_combined
         ).prediction
 
+        # Add persistence skip connection if requested
+        persistence: TensorNTCHW | None = inputs[self.output_space.name][
+            :, -1, self.target_variable_indices, :, :
+        ].unsqueeze(1)
+
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/models/encode_process_decode.py:157]: Done in b1dfe27e

**Comment by Shaerdan** [comment]
Really nice results, and i will update issue #409 to make a ref to this PR as the origin; basically the one little extra step in #409 is a simple loop back when time-stepping.

im adding these two comments here, they follow from the same experiments, implementable within this PR 

1. add an advancing anchor option 
this is already covered in the note you mentioned above: as you said right now every forecast step fuses with the same frame (day 0 observation). So the lead 14 output is a correction to a two week old anchor, and the network must synth the entire 14 day change in one shot per lead there. I think that's exactly why "evolution over 14 days is still poor", the skip fixes the baseline but not the time stepping. So I tested this ansatz in my toy example, a minimal fix:
skip_0 = persistence
for k in 1 to n_f:
      out_k  = fuse(prediction_k, skip_{k-1})
      skip_k = out_k

with additive fusion this turns the decoder to a daily integrator/corrector (ie, out_k = out_{k-1} + delta_k); with gated it becomes a learned per-pixel/per-cell mixture of 'keep yesterday' vs 'commit the predicted change"'. The extra cost is the decoder fusion runs sequentially over the time steps instead of one broadcast, but the decoder fusion is not costly compared to the processor, in the toy example test the time cost didnt change that much (i didnt rigorously time it though). do you want to try this out as another config option maybe? 

2. initialise at persistence exactly
with additive skip, the untrained model (eg epoch 0, 1st forward pass) produces persistence + decoder noise; with gated, roughly a half/half blend (sigmoid gate near 0.5 at initialisation). But if you do initialising with zeros of the decoder final convolution (additive) or biasing the gate toward the skip, it should make the untrained model reproduce persistence exactly, so training starts on the baseline and can only leave it by learning something; in my toy example test-runs this also removed the early training instability where the model first has to climb back to baseline, which is quite nice because i had that issue for a long time.

For organising, i realise that #405 and #410 changes the same files (encode_process_decode.py, base_decoder.py), i should have noticed 405 but I missed it. If it helps, I can also push the anchor option as a commit onto this branch for you to review/combine. what do you think?

**Comment by jemrobinson** [comment]
@Shaerdan: I'm happy to implement those, but perhaps it would make more sense to merge this as-is and for you to update #409 to run on top of this? Then that PR can focus on "Improve residual/skip connection functionality" and this one can be "Implement first attempt at residuals/skip connections"? I don't want to minimise the excellent work you've done there.

If you're happy with that, let me know what (if anything) from #409 should live here.

**Comment by Shaerdan** [comment]
> @Shaerdan: I'm happy to implement those, but perhaps it would make more sense to merge this as-is and for you to update #409 to run on top of this? Then that PR can focus on "Improve residual/skip connection functionality" and this one can be "Implement first attempt at residuals/skip connections"? I don't want to minimise the excellent work you've done there.
> 
> If you're happy with that, let me know what (if anything) from #409 should live here.

ok sounds good, that's the plan, and in that case nothing from #409 needs to be here.

File: icenet_mp/types/enums.py [hunk 5]
```diff
@@ -25,14 +24,11 @@ class RangeRestriction(StrEnum):
...
 
+class SkipConnectionType(StrEnum):
+    """Enum for decoder skip connection types."""
 
-class TensorDimensions(IntEnum):
-    """Enum for tensor dimensions."""
-
-    THW = 3  # Time, Height, Width
-    BTCHW = 5  # Batch, Time, Channels, Height, Width
+    ADDITIVE = "additive"
```
Comment (Shaerdan) [inline icenet_mp/types/enums.py:31]: at some point we can maybe also add concatenation skip as an option? just leaving this note here for bookmarking

**Review by Shaerdan — APPROVED** [review]
All good, LGTM
